design-decisions/2019-03-20-Ensures-for-Axioms.md

```diff
@@ -0,0 +1,122 @@
+```
+
+Claim:
+```
+If (a < b) then b else a ⇒ ∃ r . r ∧ ((a ≤ b ∧ r=b) ∨ (b ≤ a ∧ r=a))
+```
+
+Note that we do need ensure clauses for claims.
+
+Questions
+---------
+
+Should we allow ensure clauses in axioms?
+
+It is not at all obvious that we actually need them. Also, it is very easy to
+make mistakes when writing them, should we try to detect errors?
+
+As an example, the following axiom infers that `a ≤ b` out of thin air:
+```
+random(a,b) ⇒ ∃ m . m ∧ (m ≥ a ∧ m ≤ b)
+```
+
+
+Decision: Postpone allowing them
+--------------------------------
+
+We should show an error when encountering rewrite axioms with ensures. When
+needed, er should implement the decision below. We should only allow
+ensures axioms that use existential quantifiers if we don't have concrete
+examples without quantifiers.
+
+Decision: Check ensure validity
+-------------------------------
+
+We should have a flag `--rewrite-axiom-check` with three possible values:
+`None`, `BestEffort` and `Strict`. We should check the
+`⌈left⌉ ∧ requires ∧ ¬ensures` expression from the section below like this:
+
+1. `None`: no checks.
+1. `BestEffort`: check whether it is `⊥`, or the SMT gives up on checking it,
+   fail otherwise.
+1. `Strict`: check whether it is `⊥`, fail otherwise.
+
+Reasoning
+---------
+
+Let us take a `left ∧ requires ⇒ right ∧ ensures` formula (no quantification
+for now). This is equivalent to:
+```
+left ∧ requires ⇒ right ∧ ensures
+left ∧ requires → • right ∧ ensures  -- expand ⇒
+left ∧ requires → ensures ∧ • right  -- predicates can move out of •
+(left ∧ requires → ensures)          -- a → b ∧ c iff (a → b) ∧ (a → c)
+    ∧ (left ∧ requires → • right)
+```
+Now, note that this is an axiom, so the entire expression must be a
+`⊤` predicate. This means that both `left ∧ requires → • right` and
+`left ∧ requires → ensures` are `⊤` predicates.
+
+But `ensures` is a predicate.
+1. If `ensures` is `⊥`, then `left ∧ requires` is `⊥`, which means that either
+   `left` or `requires` are `⊥`, which means that `⌈left⌉` is `⊥`.
+1. If `ensures` is `⊤`, then the right hand side of the implication does not
+   matter.
+
+So we should be able to replace that part of the axiom with:
+```
+⌈left⌉ ∧ requires → ensures
+```
+Which is equivalent to
+```
+¬ (⌈left⌉ ∧ requires ∧ ¬ensures)
+```
+Which we can try to validate, or we can try to see whether
+`⌈left⌉ ∧ requires ∧ ¬ensures` is always `⊥`, i.e. whether it is unsatisfiable.
+
+Now, let us take an axiom with a quantified right hand side,
+`left ∧ requires ⇒ ∃ Z . right ∧ ensures`. Then this is equivalent to:
+```
+left ∧ requires ⇒ ∃ Z . right ∧ ensures
+left ∧ requires → • ∃ Z . right ∧ ensures  -- expand ⇒
+left ∧ requires → ∃ Z . • right ∧ ensures  -- ∃ commutes with •
+∃ Z . left ∧ requires → • right ∧ ensures  -- ∃ commutes with ∨ and →
+∃ Z . ( (left ∧ requires → • right)        -- a → b ∧ c iff (a → b) ∧ (a → c)
+      ∧ (left ∧ requires → ensures)
+      )
+```
+
```

Comment:
I think the argument below could have probably made easier, by arguing that 
`∃ Z . (A /\ B ) -> (∃ Z . A) /\ (∃ Z . B )` which means that, in particular `(∃ Z . B )` is a theorem.

Also, I don't think you can discard the forall quantifier when checking UNSAT, because when checking satisfiability all vars are existentially quantified by default.